design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+* After you run `kubectl apply -f <directory of YAMLs>`, a dependency graph of
+  resources starts to resolve. However, it's not always clear for user to know _when_
+  everything is actually ready to use unless they know readiness of which resources
+  signal the readiness of the whole environment.
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
+
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
+  translated to the YAMLs that are already deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
+  dependency so that user knows what CR instances will be deployed by the given
+  configuration stack.
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
+    the stack uses and expose the high level variations of the parameters through
+    `spec` instead of having different custom resource definitions for each class
+    of configuration.
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as `Provider` reference to refer in the deployed resource
+  or `region`.
+* Controller should signal readiness of all resources in CR's status.
+* All resources should be labelled referring to the CR instance of the stack.
+* In YAMLs, resources should refer to each other using cross-resource references
+  wherever possible.
+* **TBD** Controller should update the resources continuously and treat the given CR as
+  the source of truth even though user manually changes the resources that are
+  deployed by the stack.
+* **TBD** CRD of the configuration stack should be namespaced so that different
```

Comment:
How I imagined is like the following; the users and UI clients can create as many config CRs as they want. To prevent the name collisions of deployed cluster-scoped resources, we'd prefix the names with `namespace-name-` of the specific config CR. This way people can have many configuration environments independently in one Crossplane instance.

This thinking treats the creator of the config CR as app user rather than infra owner role that  who has cluster-wide access. Technically, implementing the CRDs namespaced or cluster-scoped requires same amount of effort. So, if the creator is assumed to be the infra owner and we don't expect app developers to create config CRs, we can go with cluster-scoped.

Reply:
+1 for cluster-scoped. I think that these stacks could be used for base-level infrastructure across multiple teams, so it seems likely they will likely be deployed by infra-owners in most cases.

Reply:
Dropping a note here that if these easy stack custom resources are namespaced we won't be able to rely on garbage collection to cleanup the cluster scoped managed resources and classes they create, thus introducing a touch more complexity to handle cleanup. I'm thinking we should assume installing an easy stack is the domain of the infrastructure operator (and thus they're cluster scoped) unless we think it's really compelling to allow app owners to install them.

Reply:
Cluster-scoped sounds good.